errors: Low level (de)ser errors

muzarski opened the pull request:
Ref: https://github.com/scylladb/scylla-rust-driver/issues/519

# Motivation

This commit is a small step forward to resolving the issues related to errors in rust driver. The main goal of this PR is to get rid of `ParseError` from `frame::types` module.

As a bonus, the newly introduced deserialization module is adjusted as well. Thanks to that, we remove the cycle in error types (`DeserializationError <-> ParseError`).

# Changes
- introduced new enum error type returned from types module: `LowLevelDeserializationError`
- changed the return error type of `types::read_*` functions to `LowLevelDeserializationError`
-  replaced `value::BuiltinDeserializationErrorKind::GenericParseError` with new variants
- changed the return error type of `deser_cql_value` to DeserializationError
- replaced `TabletsParsingError::Parse` with `TabletsParsingError::DeserializationError` 

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~[ ] I added relevant tests for new features and bug fixes.~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- ~[ ] I have adjusted the documentation in `./docs/source/`.~
- ~[ ] I added appropriate `Fixes:` annotations to PR description.~

muzarski commented:
v1.1: rebased on main

muzarski commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -52,3 +54,11 @@ pub enum ParseError {
+#[non_exhaustive]
+#[derive(Error, Debug)]
+pub enum LowLevelSerializationError {
+    #[error("Integer conversion out of range: {0}")]
+    TryFromIntError(#[from] std::num::TryFromIntError),
```
The error message is: "out of range integral type conversion attempted". I'll leave it as `error(transparent)` then, since we cant provide more context in this case.

muzarski commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -60,9 +60,7 @@ impl<'frame> DeserializeValue<'frame> for CqlValue {
         v: Option<FrameSlice<'frame>>,
     ) -> Result<Self, DeserializationError> {
         let mut val = ensure_not_null_slice::<Self>(typ, v)?;
-        let cql = deser_cql_value(typ, &mut val).map_err(|err| {
-            mk_deser_err::<Self>(typ, BuiltinDeserializationErrorKind::GenericParseError(err))
-        })?;
+        let cql = deser_cql_value(typ, &mut val).map_err(deser_error_replace_rust_name::<Self>)?;
         Ok(cql)
```
If I remove this, the following test fails:
```failures:

---- types::deserialize::row::tests::test_row_errors stdout ----
thread 'types::deserialize::row::tests::test_row_errors' panicked at scylla-cql/src/types/deserialize/row.rs:651:13:
assertion `left == right` failed
  left: "i64"
 right: "scylla_cql::frame::response::result::CqlValue"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

muzarski replied:
I set the `CqlValue` as `rust_type` in `deser_cql_value` in case of nested types (e.g. `CqlValue::Tuple`).

wprzytula replied:
OK, got it.

muzarski commented:
v2:
- addressed review comments
- converted error type for all `types::[read/write]_` to `LL(De)SerializationError`. Previously, I forgot to unify the return error types for all of the utility functions. This is helpful, so the higher level errors won't need to handle `std::io::Error/std::num::TryFromIntError` separately.

wprzytula commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -52,3 +56,48 @@ pub enum ParseError {
+    TryFromIntError(#[from] std::num::TryFromIntError),
+}
+
+/// A low level deserialization error.
+///
+/// This type of error is returned when deserialization
+/// of some primitive value fails.
+///
+/// Possible error kinds:
+/// - generic io error - reading from buffer failed
+/// - out of range integer conversion
+/// - conversion errors - e.g. slice-to-array or primitive-to-enum
+/// - not enough bytes in the buffer to deserialize a value
+#[non_exhaustive]
+#[derive(Error, Debug)]
+pub enum LowLevelDeserializationError {
```
Now I'm very happy with the docstrings :)

wprzytula commented:
What about adding context to LLDeserError in new deserialization framework?

Lorak-mmk commented on scylla-cql/src/frame/types.rs:
```diff
@@ -161,20 +162,22 @@ impl<'a> RawValue<'a> {
     }
     let (ret, rest) = buf.split_at(count);
     *buf = rest;
     Ok(ret)
 }
 
-pub fn read_int(buf: &mut &[u8]) -> Result<i32, std::io::Error> {
+pub fn read_int(buf: &mut &[u8]) -> Result<i32, LowLevelDeserializationError> {
```
This makes returned type less specific. If this function can only return io error then why force the caller to handle other errors?

Lorak-mmk commented on scylla-cql/src/frame/types.rs:
```diff
@@ -207,7 +210,7 @@ fn type_int() {
     }
 }
 
-pub fn read_long(buf: &mut &[u8]) -> Result<i64, std::io::Error> {
+pub fn read_long(buf: &mut &[u8]) -> Result<i64, LowLevelDeserializationError> {
```
This applies also here and other places where `std::io::Error` was replaced

Lorak-mmk commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -1656,10 +1651,18 @@ impl Display for BuiltinDeserializationErrorKind {
     }
 }
 
+impl From<LowLevelDeserializationError> for BuiltinDeserializationErrorKind {
+    #[inline]
+    fn from(err: LowLevelDeserializationError) -> Self {
+        Self::LowLevelDeserializationError(err)
+    }
+}
+
```
In other places we are using `thiserror` crate to generate the conversions. Why not do the same here?

Lorak-mmk commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -37,6 +37,8 @@ pub enum FrameError {
 pub enum ParseError {
```
Please remind me, is the current plan to get rid of ParseError completely?
If not, we need to remember to add non_exhaustive to it, and other errors (like FrameError above).

muzarski replied:
I ultimately would like to get rid of `ParseError`. It will be decoupled into multiple error types, which will be more specific and what's the most important - they will be typed. As of now, most of the `ParseError`s are stringified (e.g. converted to `QueryError::InvalidMessage`).

There is a similar issue with a `FrameError`, however I'm not planning to remove it. I'd like to decouple it and introduce a distinction between `FrameSerializationError` (for client frames) and `FrameDeserializationError` (for server frames). These two could be then represented as variants of `{Query/NewSession}Error` (and not stringified to `QueryError::InvalidMessage`, which is something we currently do).

Lorak-mmk reviewed (changes requested):
I love it when github decides to post my review but without the main comment I wrote :) Posting again...

I think those are good changes in the right direction in general.

Please explain, because I don't remember from our discussion and I can't understand from commits: why do we have a distinction between deserialization error and low level deserialization error? The errors are a public API, why should a user of a library care which internal layer of the driver returned the error and not just what the error is?


Note: If we don't have something like this, whole error refactor should be finished with tests that will present and assert the error messages in various conditions. That way we can see them without writing one-off code to trigger those errors.

wprzytula commented:
> Please explain, because I don't remember from our discussion and I can't understand from commits: why do we have a distinction between deserialization error and low level deserialization error? The errors are a public API, why should a user of a library care which internal layer of the driver returned the error and not just what the error is?

LowLevelDeserializationError is strictly connected to the CQL protocol types deserialization layer. On top of it, there is DeserializationError, which carries the meaning of a logical error during deserialization.
For example, the underlying CQL protocol type structure can be perfectly valid (so no LowLevelDeserializationError involved), but invalid for the expected logical structure of, say, a CQL Set. (e.g., an `int` present when `bytes` or `string` was expected). While the example is not perfect, I hope you get the difference.

muzarski commented:
v3:
- rebased on main
- removed `LowLevelSerializationError` at all. The reason is that `types::write_*` functions all return `std::num::TryFromIntError`, so currently there is no need to wrap it with an additional error type.
- narrowed the return type of `types::read_*` functions to `std::io::Error` (if possible)
- added more context to deserialization framework. Introduced three new variants which wrap `LowLevelDeserializationError`:
  - BadDate - deserialization of some date's field failed (months, days, or nanoseconds)
  - BadDecimalScale - deserialization of decimal's scale failed
  - RawCqlBytesReadError - failed to read raw cql value bytes from a frame slice

Lorak-mmk commented:
>     * removed `LowLevelSerializationError` at all. The reason is that `types::write_*` functions all return `std::num::TryFromIntError`, so currently there is no need to wrap it with an additional error type.
> 

I didn't read the changes yet, so my question may be inaccurate.
Isn't it possible for this situation to change in the future?
If it is, then `LowLevelSerializationError` is better because we can add variants to it.
If it can't then nvm
